### ISSUE TITLE
Support HDR image formats (like JXR)

### ISSUE THREAD

**Issue by fofwisdom** [pixelfed/pixelfed#4486]
Windows 11 supports JXR and produces a native JXR. It has quickly become popular. It would be great to have a roadmap for JXR support in Pixelfed.

**Comment by trwnh** [comment]
JPEG XR is not "popular"; it is also no longer supported and has been superseded by JPEG XL.

**Comment by fofwisdom** [comment]
> JPEG XR is not "popular"; it is also no longer supported and has been superseded by JPEG XL.

I'm surprised. 
Thanks for the feedback and advice.
Microsoft & Nvidia has been coding for JXR support for recently years. All Windows 11 users get JXR when they press to capture a screen in HDR situations.
Does this mean that Microsoft will also be switching support to JXL?

**Comment by fofwisdom** [comment]
JXL has also been abandoned, so we'll have to wait for WebP 2. HDR will never see the light of day in the web world while JXR stacks up locally on Windows users for years to come.
Too bad.
Thanks.

**Comment by trwnh** [comment]
I'm not sure what Microsoft will be doing, but I guess for Pixelfed the best we can do is maybe see if it can be converted by gd or imagick. But I suspect if it's already not working, then they likely don't support it. There's also a config option for which file types are allowed to be uploaded, but it's what happens after upload that matters.

I can reopen this issue as a more general request for HDR image support, but I suspect that we will be waiting for a while until the tooling supports it... Of the HDR formats listed in https://en.wikipedia.org/wiki/Category:High_dynamic_range_file_formats I see AVIF and HEIF as the only ones that are likely to be supported. The bottleneck is in the image processor running on the Pixelfed host server -- putting aside GD which only supports some basic formats, ImageMagick supports a lot of formats but not every format: https://imagemagick.org/script/formats.php#supported

In theory JXR is listed there but it depends on `jxrlib`. So it is possible that it could work, but only under the following requirements:

- The admin has configured the server to use `imagick` instead of `gd`
- The admin has installed `jxrlib` as an extra dependency
- The admin has allowed upload of `image/vnd.ms-photo, image/jxr` MIME types

after which, the image would probably be converted to a normal JPEG for web viewing, and in doing so it loses HDR information probably?

**Comment by fofwisdom** [comment]
My thoughts were short.

It seems like the market and the community are going their own way without reaching a consensus. There is also the issue of the direction of image standards by web browser engine maintainers.

When converting HDR to SDR, the information from the HDR tonemap is overwritten into the image with brightness adjustments that are fixed and locked in.

If Pixelfed were to apply this fixed value, it would be confusing for users and could lead to difficulties for maintainers and users alike.

Native HDR or conversion support in web browsers is still a challenge.

Thank you for your consideration.

**Comment by lipupini** [comment]
So with this JXR format, the HDR tonemap is stored within the image itself

In theory this would allow adjustments to the HDR setting on the client side

If you are showing an image in a gallery or portfolio, isn't that typically the final edited version? I'm curious as to the primary use of being able to store the HDR information. Is this mostly to eliminate the typical step of converting to a standard presentation format for uploading?

**Comment by Saltane** [comment]
> JXL has also been abandoned…

I am not sure what you mean by that, the format is being in active development and got itʼs last commit 3 days ago (https://github.com/libjxl/libjxl). Itʼs still very promising format with only one single company being in favor of WebP2 — Google, the creators of WebP2.

+1 for JXL support.

**Comment by Openhuman2** [comment]
+1 for jxl support now that google is reimplementing it. 
It has many benefits beeing a one file format for nearly everything but smaller.
(png jpeg tiff heif raw webp) Basically I personally transcode everything now to jxl.
also it can load much faster and give a nearly instant quick preview of the picture.

see the official site for further information.
google will re-release jxl support approximately by march 2026

**Comment by fofwisdom** [comment]
> I'm curious as to the primary use of being able to store the HDR information.

HDR is no longer just for professional final editing content. In Windows apps like web browsing and gaming, wide color gamut can only be achieved in an HDR environment. And the native screenshots of applications consumed in these environments are JXR. General consumers worldwide have been using JXR for a long time.

The problem was that converting JXR to JPEG caused lossy issues, so many consumers were waiting for HDR image format support.

However, with the Windows 25H2 update and the Photos app update, conversion to JPEG is now possible.

Therefore, it seems the demand for JXR format support among regular users will likely decrease again.

**Comment by dshatz** [comment]
Hi I'd like to help implementing this. Can someone give a hint about where to start? Is this on the client or somewhere deep in the protocol?

Edit:
Assuming it is here: https://github.com/pixelfed/pixelfed/blob/88950a10e25987bdbb7b6fc555b09e52e9c577ec/app/Util/Media/Image.php#L117

Any comments?

**Comment by gregbenz** [comment]
@dshatz Here is information for developers on the most important HDR formats: https://gregbenzphotography.com/hdr/#developers

The key formats are **JPG with an ISO gain maps**, and soon **AVIF with an ISO gain map** (much smaller, higher quality). The gain maps are key to offering highest quality on less capable displays. The open source Google library libultrahdr supports JPG now and their official roadmap lists AVIF and HEIF. That library is the basis of support for many other libraries.
 
Other HDR formats which are likely to be used: AVIF (without a gain map) is used now, and longer term JXL with or without a map (JXL support is only Safari for now, but Chrome has support in Canary under a dev flag).

There are some uses of PNG for HDR, but I wouldn't expect it grows much but I believe is used for screenshots on Android and I don't believe the encoding uses ISO standards. There are older JPGs with proprietary XML encoding (three different approaches to this XML), but ISO has been well supported for a while now.

You can use this page to help compare source to any transcoded image (see the video to understand the key issues it helps expose when transcoding HDR): https://gregbenzphotography.com/hdr-soft-proofing-test-hdr-gain-maps/

**Comment by dshatz** [comment]
After some research and testing, I think it makes sense to adopt UltraHDR jpgs as the storage format to maximize compatibility.

1. Can be displayed anywhere including super ancient devices.
2. Looks good on SDR displays.

So, pixelfed could be accepting uploads in various formats and converting them to UHDR jpegs as part of `ImageOptimize` job. Using [libvips](https://github.com/libvips/libvips) seems to be the right choice here.
 - UltraHDR Jpg - no conversion needed
 - JPG with Apple gainmaps - depends on [PR in libultrahdr](https://github.com/google/libultrahdr/pull/385) **(merged)**
 - 10 bit HDR HEIC, AVIF, JXL (no gainmap) - depends on PRs in libvips [1](https://github.com/libvips/libvips/pull/4996), [2](https://github.com/libvips/libvips/pull/5101) **(merged)**
 - JPG with ISO gainmap - [Supported by ultrahdr](https://github.com/google/libultrahdr/blob/main/CMakeLists.txt#L137)
 - HEIC/AVIF with ISO gainmap - will be supported 

Would this approach be desirable?

EDIT:
Anyone can provide samples for JPG, HEIC, AVIF with ISO gainmaps?

**Comment by gregbenz** [comment]
Ultra HDR is more of a marketing term and could be vague from a standards perspective - it often refers to the Android XMP encoding spec. That spec was created before there was an official ISO standard. It also only applies to JPG (there is no similar XMP format for AVIF, JXL, etc). While it is fairly well supported, it is not the way forward and adds ~2K to the file size. It is probably best at this time to not output with this format. (note that there is a very similar Adobe XMP spec and an undocumented Apple spec, neither of these should be exported and are not supported by libultra).

What you really need is ISO-standard encoding. It is very well supported and is the offical spec. It's metadata is compact and written in the binary codestream (not XMP).

Libultra supports encoding both ISO and the Android XMP encodings. It can also dual encode to support both (these are configurable options when compiling). I recommend outputting ISO only as it is very well supported and avoids the file bloat when including the redundant / deprecated XMP.

To compile for ISO only, you would compile libultrahdr using:
cmake **-DUHDR_WRITE_ISO=1** **-DUHDR_WRITE_XMP=0** -G \"Unix Makefiles\" ../

libvips uses libultrahdr as a dependency.

Sample ISO encodings for gain maps for JPG, AVIF, and JXL are available at https://helpx.adobe.com/camera-raw/using/gain-map.html#resources

Note that gain maps may be encoded with the base image as either the SDR or HDR (not applicable for JPG as 8-bits is insufficient for a standalone HDR). Adobe has both of these. Either approach is well supported for AVIF outside of FireFox (which does not understand gain maps nor HDR). But encoding with an HDR base is ideal as it allows more aggressive compression (since the SDR rendition is less important). It's probably also ideal for a better experience when downloading on a slow connection (to minimize jumps when the base is decoded before the map segment is fully downloaded). The key should be to preserve what was in the source, and libultra should mange this when supported - I'm just noting an important test case to consider for AVIF.

**Comment by dshatz** [comment]
> What you really need is ISO-standard encoding. It is very well supported and is the offical spec. It's metadata is compact and written in the binary codestream (not XMP).

Right that would be desirable but I would like to bring to attention that currently the industry standard is effectively what libultrahdr does by default (i.e. Write XMP only I think, but it seems to be changing soon). **I don't expect pixelfed hosters to recompile libultrahdr (and vips).**

Basically to clarify my previous comment - storage format should be an SDR image (jpeg/avif) with a gainmap for better compatibility. The gainmap can be XMP while libultrahdr is doing that and ISO afterwards. Both are supported well for display.

**Comment by gregbenz** [comment]
If the gain map lacks ISO encoding, that would be a bad result (having dual
encoding with XMP is safe). I can’t recall libultra’s default compiler
setting (last mainline release was quite a while ago and ISO may not have
been enabled by default, but it works very well and should be enabled).

An XMP-only export should be considered a failed result in testing /
validation. That will not be universally supported going forward as is not
compliant with the official standard.

**Comment by dshatz** [comment]
> ISO may not have
been enabled by default, but it works very well and should be enabled

I agree it should be enabled. Hopefully it will.


> That will not be universally supported going forward

I haven't seen any device where XMP gainmap was not decoded, sorry.
Going forward ISO is the standard. **Today** libultrahdr shipped by distros does not output that. 

Let's wait for the mentioned PRs ([comment](https://github.com/pixelfed/pixelfed/issues/4486#issuecomment-4064279465)) to be merged, as well as [this](https://github.com/ittiam-systems/libultrahdr/tree/experimental).
Once that happens, we can easily convert from anything to JPEGs/AVIFs with a gainmap.

**Comment by gregbenz** [comment]
It is unclear when those libultrahdr updates will be made, the experimental branch has not been updated in quite some time. I would recommend not waiting for it (it generally works well, but the AVIF output from the experimental is either not uniform or there is a bug in Safari as it does not show HDR there). 

The proposed AVIF/HEIF updates should not be breaking changes for JPG support now, and the project does not offer pre-built binaries so you'd be compiling for ISO support anyway. Though it's a moot point if you're using libvips as you would not be ingesting libultra directly. Same for Sharp, which is based on libvips.
